python/data_structures/least_recently_used_cache.py

```diff
@@ -0,0 +1,61 @@
+# Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and put.
+
+# get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
+# put(key, value) - put or insert the value if the key is not already present. When the cache reaches its capacity, it should invalidate the least recently used item before inserting the new item.
+# The LRUCache will be initialized with an integer corresponding to its capacity. Capacity indicates the maximum number of unique keys it can hold at a time.
+
+# Definition of “least recently used” : An access to an item is defined as a get or a put operation of the item. “Least recently used” item is the one with the oldest access time.
+
+#  NOTE: If you are using any global variables, make sure to clear them in the constructor.
+
+#  Input :
+#          capacity = 2
+#          put(1, 10)
+#          put(5, 12)
+#          get(5)        returns 12
+#          get(1)        returns 10
+#          get(10)       returns -1
+#          put(6, 14)    this pushes out key = 5 as LRU is full.
+#          get(5)        returns -1
+
+
+class LRU:
+    def __init__(self,capacity):
+        self.capacity = capacity
+        self.cache = {}
+        self.lru = []
```

Comment:
Is this an array you're using here? Doesn't that mean that every `reorder` (and, incidentally, `get`) will be Θ(n)?

Reply:
This comment got me thinking, why not use an `OrderedDict` that can function in place of both `cache` and the `lru` list?

```
d = OrderedDict()
d[1] = 10
d[5] = 12
# when you want to push the third key (out of capacity)
del d[d.keys()[0]]
d[6] = 14
```

Or was the intention not to use anything from `collections`?

Reply:
What is an ordered dict? A hashMap whose keys are sorted? :thinking:

Reply:
Oui

Reply:
`Ordered dictionaries are just like regular dictionaries but they remember the order that items were inserted. When iterating over an ordered dictionary, the items are returned in the order their keys were first added.`
https://docs.python.org/2/library/collections.html#collections.OrderedDict

Reply:
If you define an ordinary dictionary in python and call `keys()` on it, it can return the keys in an arbitrary order

Reply:
@varunganesh91 Ooo interesting, didn't know of the existence of `OrderedDict()`, and yes, I was trying to use simple data structures to write the LRU cache.

@Bertrand31 Hmm yea, every reorder will be Θ(n)... A PriorityQueue would have provided better time complexity

Reply:
Using stuff from collections is a nice idea.

I tried implementing the LRUCache with a `deque` and a `ChainMap`.

```
class LRUCache(object):
     def __init__(self, cache_size):
         self.cache = deque(maxlen=cache_size)
     def set(self, key, value):
         self.cache.appendleft({key: value})
     def get(self, key):
         cached_value = ChainMap(*self.cache).get(key, -1)
         if cached_value != -1:
             self.cache.rotate(len(self.cache) - self.cache.index({key: cached_value}))
         return cached_value
```
```
>>> a = LRUCache(cache_size=2)
>>> a.set(1, 10)
>>> a.set(5, 12)
>>> a.get(5)
12
>>> a.cache
deque([{5: 12}, {1: 10}], maxlen=2)
>>> a.get(1)
10
>>> a.cache
deque([{1: 10}, {5: 12}], maxlen=2)
>>> a.get(10)
-1
>>> a.cache
deque([{1: 10}, {5: 12}], maxlen=2)
>>> a.set(6, 14)
>>> a.cache
deque([{6: 14}, {1: 10}], maxlen=2)